pkg/querier/blocks_store_querable_streaming.go

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querier
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storegateway/storegatewaypb"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+	"github.com/grafana/mimir/pkg/util/limiter"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// StreamingSeries holds the labels of the streaming series and the source to get the chunks
+// for the series.
+type StreamingSeries struct {
+	Labels []mimirpb.LabelAdapter
+	Source StreamingSeriesSource
+}
+
+// StreamingSeriesSource holds the relationship between a stream of chunks from a SeriesChunksStreamReader
+// and the expected position of a series' chunks in that stream.
+type StreamingSeriesSource struct {
+	StreamReader *SeriesChunksStreamReader
+	SeriesIndex  uint64
+}
+
+// SeriesChunksStreamReader is responsible for managing the streaming of chunks from a storegateway and buffering
+// chunks in memory until they are consumed by the PromQL engine.
+type SeriesChunksStreamReader struct {
```

Comment:
Good idea. Ingester code is inside `pkg/ingester/client`. It can be brought into `pkg/ingester`. I will move the code from this PR into `pkg/storegateway`.

Reply:
Turns out I cannot put it in `pkg/ingester` for the ingester streaming because of some cyclic dependency.